### PR TITLE
Add header models generation

### PR DESCRIPTION
**Pull request by pakrym**
This is done on top of https://github.com/Azure/autorest.csharp/pull/336 so might include more changes and needs to be rebased.

### CODE REVIEW

File: src/AutoRest.CSharp.V3/CodeGen/CodeWriter.cs [hunk 1]
```diff
@@ -202,6 +206,10 @@ public CodeWriter Append(string str)
 
         public CodeWriter Comma() => Append(", ");
 
+        public CodeWriter Space() => Append(" ");
+
+        public CodeWriter Semicolon() => Append(";").Line();
```
Comment (MiYanni) [inline src/AutoRest.CSharp.V3/CodeGen/CodeWriter.cs]: Maybe `SemicolonLine` would be a better name.

File: src/AutoRest.CSharp.V3/CodeGen/ClientWriter.cs [hunk 2]
```diff
@@ -344,24 +345,53 @@ private void WriteQueryParameter(CodeWriter writer, QueryParameter queryParamete
...
                     writer.Line($"case {statusCode}:");
                 }
 
-                writer.Append($"return {writer.Type(typeof(Response))}.FromValue(");
-                writer.ToDeserializeCall(schemaResponse!, _typeFactory, "document.RootElement", writer.Type(responseType), responseType.Name ?? "[NO TYPE NAME]");
-                writer.Line(", response);");
+                bool hasBody = bodyType != null;
+                using (hasBody ? writer.Scope() : default)
+                {
+                    if (hasBody)
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AutoRest.CSharp.V3/CodeGen/ClientWriter.cs]: let me check if we still need that.

File: src/AutoRest.CSharp.V3/CodeGen/CodeWriter.cs (hunk 1 above)
Comment (pakrym) [reply src/AutoRest.CSharp.V3/CodeGen/CodeWriter.cs]: My idea was that there is always a line after semicolon

File: src/AutoRest.CSharp.V3/ClientModels/ClientBuilder.cs [hunk 3]
```diff
@@ -71,9 +73,11 @@ internal class ClientBuilder
                             CreateDefaultValueConstant(requestParameter), false);
                         break;
                     default:
-                        constantOrParameter = new ServiceClientMethodParameter(requestParameter.CSharpName(),
+                        constantOrParameter = new ServiceClientMethodParameter(
+                            requestParameter.CSharpName(),
                             ClientModelBuilderHelpers.CreateType(requestParameter.Schema, requestParameter.IsNullable()),
-                            CreateDefaultValueConstant(requestParameter), requestParameter.Required == true);
+                            CreateDefaultValueConstant(requestParameter),
+                            requestParameter.Required == true);
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AutoRest.CSharp.V3/ClientModels/ClientBuilder.cs]: It is, requestParameter.Required is nullable while the parameter isn't

File: src/AutoRest.CSharp.V3/CodeGen/ClientWriter.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AutoRest.CSharp.V3/CodeGen/ClientWriter.cs]: meh, I'll just use !=

File: src/AutoRest.CSharp.V3/CodeGen/CodeWriter.cs (hunk 1 above)
Comment (MiYanni) [reply src/AutoRest.CSharp.V3/CodeGen/CodeWriter.cs]: Not if a semicolon is used in a for loop. 🤸‍♂

**Pull request merged**